two_factor/views/core.py

```diff
@@ -159,29 +175,36 @@ def get_user(self):
         Returns the user authenticated by the AuthenticationForm. Returns False
         if not a valid user; see also issue #65.
         """
-        if not self.user_cache:
-            form_obj = self.get_form(step='auth',
-                                     data=self.storage.get_step_data('auth'))
-            self.user_cache = form_obj.is_valid() and form_obj.user_cache
         return self.user_cache
 
-    def get_context_data(self, form, **kwargs):
+    def get_context_data(self, step='auth', user=None, form=None, **kwargs):
         """
         Adds user's default and backup OTP devices to the context.
         """
-        context = super(LoginView, self).get_context_data(form, **kwargs)
-        if self.steps.current == 'token':
-            context['device'] = self.get_device()
-            context['other_devices'] = [
-                phone for phone in backup_phones(self.get_user())
-                if phone != self.get_device()]
-            try:
-                context['backup_tokens'] = self.get_user().staticdevice_set\
-                    .get(name='backup').token_set.count()
-            except StaticDevice.DoesNotExist:
-                context['backup_tokens'] = 0
-
-        context['cancel_url'] = resolve_url(settings.LOGOUT_URL)
+        context = super(LoginView, self).get_context_data(**kwargs)
+        self.step_cache = step
+        context['step'] = step
+        if step == 'auth':
+            context['form'] = form or self.form_dict['auth']()
+        elif step in ['token', 'backup']:
+            context['username'] = self._username_cache
+            context['password'] = self._password_cache
```

Comment:
@moggers87 You're right. The auth is done at every step of the workflow. That's why we need to save the username/password in some way, even if I'm not a big fan of it either. With my fix, at least I don't save it in the DB anymore, it's only sent back to the user as hidden form input, to make the user send it again with the token form seamlessly. If you have a better idea, tell me :)